### PR TITLE
Initial implementation of virtual products

### LINKED ISSUE #349: Adding Virtual Products module to datacube
## Background

Many common use cases of Datacube (DC) involve combining temporally and spatially similar sets of data to produce output data. An example is using Product Quality (PQ) products with other products to perform cloud masking. Issues have been encountered with dataset merging previously in Datacube stats (see https://gist.github.com/Kirill888/a4b52d0077fa4e36b351c22827782492) and in Datacube WMS there are times where only TIRS sensor datasets are available, causing unexpected results. As such the creation of Virtual Products (VPs) has been considered as a way to provide users a way to easily define common patterns used in Datacube work and to solve issues seen in software which uses Datacube surrounding the merging and transformation of data without having to create bespoke code each time.

## Products in Datacube
In the current version of DC, products are unique strings which associate that string (the product name) with metadata about Datasets. This metadata includes information such as the measurements that the product contains, the grid specification and the mapping of raw fields in the dataset to computed fields (e.g. from_dt and to_dt which product a time measurement). In DC the products are represented by a `DatasetType` which has a `MetadataType`. Each `Dataset` has a `DatasetType` and the `DatasetType` and `MetadataType` are typically loaded from JSON documents stored in a database. See also (http://damien-agdc.readthedocs.io/en/extradocs/dev/data_model.html)

## Preliminary Plan for Virtual Products

A class implementing Virtual Products (VPs) will represent a tree which represents the hierarchy of virtual products. Leaf nodes will be Virtual Products which are unmodified products.

### Combinators
Combinators are functions which accept 1 or more Virtual Products and return a Virtual Product. The following combinators will be available:

- `Collate : List A -> A`
   For example: Combining the sensor readings for LS5 on 05/07/95, LS6 on 06/11/95, LS8 on 07/11/95 into one dataset.

- `Juxtapose: A -> B -> A x B`
   Similar to an `INNER JOIN` in SQL. This could be used for apply PQ for cloud masking in combination with `Transform`.

- `Transform`: `A -> B`
   Synonymous with the `map` function of the `MapReduce` paradigm. Will require some transformation function which accepts a dataset and returns a dataset. This combinator may also modify the type of the dataset.

### Methods & Workflow (High Level API)

- `construct`
  Constructing a Virtual Product will set it's child Virtual Products as well as any functions it requires to operate. The construction will not access the database.

- `validate`
  The validation function will check the measurement types of the inputs and outputs and ensure that they are compatible. In most cases this will be checking that they match, however with a `Transform` there may be a change.

- `query`
  The query function will retrieve datasets which match a query, product. The product being queried may be virtual. Access to the database will cease after this stage.

- `serialize` (Optional?)
   The serialize function will use the results of the `query` function (i.e. datasets that match the query) and serialize them for use on machines or nodes that cannot or should not query the database but can fetch data.

- `fetch`
  The fetch function will use the results of the `query`, direct or deserialized, and load the datasets. At this stage combinators working on data can apply their functions or predicates. Once completed, the output for the top level Virtual Product will be the desired Virtual Dataset.

### Virtual Product Module

This approach would not modify any opendatacube-core code. Instead a new module would be created that defined `VirtualProduct`. This module would make use of the `datacube` module and API.

The module would query the properties of the `DatasetTypes` (a.k.a. products) used in the definition of the `VirtualProduct` and conduct type and sanity checking on combinators. After passing the checks the storage location of datasets for the given query (e.g. geoboxed or time bounded), the database would be queried using `datacube`. Advanced versions could intelligently walk the product tree to determine how to construct a SQL query for the various products and / or cache query results.

The module would then load the datasets from their location (e.g. disk, aws s3, etc.). Once the datasets are loaded, combinators are applied to the datasets and the results until the final result is created and returned. Again intelligent caching of result could be applied in this stage.

### Potential ideas / constraints for features (in progress)

- Metadata as bands, e.g. data about what LandSat data came from may be encoded as metadata which will be preserved as a band
- Timestamps must be retained
- Must support DC GridWorkflow for datasets which will not fit into memory
- For all queries against a virtual product the database should be accessed to retrieve the locations of datasets, and then database should not be accessed again. The locations of the datasets may need to be serialized to a storage device and loaded later.
- Support querying multiple databases.

### PR DESCRIPTION
**Pull request by uchchwhash**
### Reason for this pull request
The motivation is detailed in #349, [here](https://github.com/orgs/opendatacube/teams/developers/discussions/4), and [here](https://github.com/orgs/opendatacube/teams/developers/discussions/3).

Also related #194.

### Proposed changes
- Add virtual products (`collate`, `juxtapose` and `transform` datacube products)
- Add common transformations of data (`MakeMask`, `ApplyMask`, `ToFloat`, and `Rename` to start with)
- Add support for constructing virtual products from (`yaml`) recipe

 - [ ] Closes #349

### CODE REVIEW

File: datacube/virtual/impl.py [hunk 1]
```diff
@@ -0,0 +1,550 @@
...
+"""
+from abc import ABC, abstractmethod
+from functools import reduce
+import warnings
+
+import xarray
+import numpy
+
+from datacube import Datacube
+from datacube.compat import string_types
```
Comment (Kirill888) [inline datacube/virtual/impl.py]: unnecessary import

File: datacube/virtual/__init__.py [hunk 2]
```diff
@@ -0,0 +1,5 @@
+from .impl import BasicProduct as basic_product
+from .impl import Transform as transform
+from .impl import Transformation
+from .impl import Collate as collate
+from .impl import Juxtapose as juxtapose
```
Comment (Kirill888) [inline datacube/virtual/__init__.py]: Add `__all__ = ...` statement.

File: datacube/virtual/transformations.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
+# TODO: nice repr for the built-in transformations?
+
+import numpy
```
Comment (Kirill888) [inline datacube/virtual/transformations.py]: unnecessary import

File: datacube/virtual/impl.py [hunk 4]
```diff
@@ -0,0 +1,550 @@
...
+                                  use_threads=merged.get('use_threads'))
+
+    def __repr__(self):
+        settings = ",".join("{}={}".format(key, value) for key, value in self.settings.items())
+        return "BasicProduct(product={}, {})".format(self.product, settings)
+
+
+class Transformation(ABC):
+    @abstractmethod
+    def measurements(self, input_measurements):
```
Comment (Kirill888) [inline datacube/virtual/impl.py:109]: Needs more docs

- Description of input output data structure
- Expected error handling behaviours for implementations (exception vs return None, that sort of stuff)

File: datacube/virtual/impl.py [hunk 5]
```diff
@@ -0,0 +1,550 @@
...
+        """
+        Returns the list of output measurements from this transformation.
+        """
+        pass
+
+    @abstractmethod
+    def compute(self, data):
```
Comment (Kirill888) [inline datacube/virtual/impl.py:118]: Needs more docs about the relationship between `data` and `input_measurements` from the other method, I assume they are related?

**Review by Kirill888 — DISMISSED** [review]
Overall code looks clean, and it's not a lot of it, which is nice, but without tests I can't properly review it or merge it. Without sample code (ideally in tests) it's hard to bootstrap review process, to understand data flows and structures. In particular it's not clear what error conditions are/can be and whose responsibility they are. It's probably worthwhile creating some exception classes for common problems, wrong "type/shape" for input measurements, that sort of stuff, along with clear instructions for virtual product writers when to raise what error/return None.

Really need to think about how to test this module somehow. How much of it requires DB/real data IO, what can be tested without that.

File: datacube/virtual/__init__.py (hunk 2 above)
Comment (uchchwhash) [reply datacube/virtual/__init__.py]: Added.

File: datacube/virtual/impl.py (hunk 5 above)
Comment (uchchwhash) [reply datacube/virtual/impl.py:118]: Added docs.

File: datacube/virtual/impl.py (hunk 1 above)
Comment (uchchwhash) [reply datacube/virtual/impl.py]: Removed.

File: datacube/virtual/transformations.py (hunk 3 above)
Comment (uchchwhash) [reply datacube/virtual/transformations.py]: Removed.

**Review by GypsyBojangles — APPROVED** [review]

**Pull request merged**